### PR TITLE
Fix focus scope with changing children

### PR DESCRIPTION
**Pull request by snowystinger**
Closes https://github.com/adobe/react-spectrum/issues/2415

Keep focus scope up to date with changing children and auto focus children.
The two cases observed are:
1) Open a dialog, change the children in the dialog, hit esc, focus is lost.
This was fixed by using scopeRef.current as pointed out in the issue description because the scope was out of date.
2) Open a dialog where a child has autofocus, hit esc, focus is lost.
This was fixed by changing when nodeToRestore is set so that it runs before autoFocus takes place.

We are preferring this PR over https://github.com/adobe/react-spectrum/pull/2416 right now because it's difficult to review due to the refactor. Not that the refactor is bad, but by breaking this up, we can make sure we know what the fix is, and we can refactor it separately such that no changes to behavior occur.

Thank you @kherock for your hard work on this!

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Test steps (case 1):
1. Go to https://reactspectrum.blob.core.windows.net/reactspectrum/f976e1d87682ba3a61027506d360b91d8976e267/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain
2. Tab to "Open dialog" and hit enter. A new FocusScope row should appear and focus should be moved to the first input element. 
3. Tab to "replace focusscope children" and hit Enter. The children in the 2nd row should be replaced by 3 inputs and focus should move to the middle input element
4. Hit Escape key. The 2nd "dialog" row should close and focus should be moved to the trigger child

Case 2 is covered by the unit test

## 🧢 Your Project:

RSP

### CODE REVIEW

**Review by LFDanLu — APPROVED** [review]
LGTM

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx [hunk 1]
```diff
@@ -46,35 +47,47 @@ function MaybePortal({children, usePortal}: { children: ReactNode, usePortal: bo
...
+            <button type="button" onClick={onClose}>
+              close
+            </button>
+            {open && <NestedDialog contain={contain} onClose={() => setOpen(false)} usePortal={usePortal} />}
+          </div>
+        )}
+        {showNew && (
+          <div role="dialog" onKeyDown={onKeyDown}>
+            <input />
+            <input autoFocus />
```
Comment (majornista) [inline packages/@react-aria/focus/stories/FocusScope.stories.tsx:81]: `autoFocus` on this `input` means that focus will be maintained within the FocusScope when the children change, however, without the `autoFocus` prop, focus will be lost to the `document.body`. https://github.com/adobe/react-spectrum/pull/2763 has been opened to account for this and keep focus within the FocusScope when `contain` is `true`.

**Review by majornista — APPROVED** [review]

File: packages/@react-aria/focus/test/FocusScope.test.js [hunk 2]
```diff
@@ -331,7 +331,134 @@ describe('FocusScope', function () {
...
+      act(() => {outside.focus();});
+
+      rerender(<Test show />);
+
+      let input2 = getByTestId('input2');
+      expect(document.activeElement).toBe(input2);
+
+      rerender(<Test />);
+
+      expect(document.activeElement).toBe(outside);
```
Comment (ktabors) [inline packages/@react-aria/focus/test/FocusScope.test.js:362]: Do we care about proving the focus scope doesn't goto `.outside` when it isn't focused before we rerender it twice?

**Review by ktabors — COMMENTED** [review]
This looks good, I just have the one question in tests.

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx (hunk 1 above)
Comment (snowystinger) [reply packages/@react-aria/focus/stories/FocusScope.stories.tsx:81]: yep, #2763 is the reason we aren't addressing this issue here

**Review by ktabors — APPROVED** [review]

**Pull request merged**